### PR TITLE
:warning: Remove rpm install method

### PR DESCRIPTION
**Pull request by elfosardo**
Closes https://github.com/metal3-io/ironic-image/issues/583

### CODE REVIEW

File: prepare-image.sh [hunk 1]
```diff
@@ -2,75 +2,58 @@
...
-    if [[ ! -s "${UPPER_CONSTRAINTS_PATH}" ]]; then
-        UPPER_CONSTRAINTS_PATH="/tmp/upper-constraints.txt"
-        curl -L https://releases.openstack.org/constraints/upper/master -o "${UPPER_CONSTRAINTS_PATH}"
-    fi
-
-    if [[ -n ${SUSHY_SOURCE:-} ]]; then
-        sed -i '/^sushy===/d' "$UPPER_CONSTRAINTS_PATH"
-    fi
+if [[ -n ${SUSHY_SOURCE:-} ]]; then
+    sed -i '/^sushy===/d' "$UPPER_CONSTRAINTS_PATH"
```
Comment (tuminoid) [inline prepare-image.sh:40]:
```suggestion
    sed -i '/^sushy===/d' "${UPPER_CONSTRAINTS_PATH}"
```

File: prepare-image.sh [hunk 2]
```diff
@@ -2,75 +2,58 @@
...
+fi
 
-    if [[ -n ${IRONIC_LIB_SOURCE:-} ]]; then
-        sed -i '/^ironic-lib===/d' "$UPPER_CONSTRAINTS_PATH"
-    fi
+if [[ -n ${IRONIC_LIB_SOURCE:-} ]]; then
+    sed -i '/^ironic-lib===/d' "$UPPER_CONSTRAINTS_PATH"
```
Comment (tuminoid) [inline prepare-image.sh:44]:
```suggestion
    sed -i '/^ironic-lib===/d' "${UPPER_CONSTRAINTS_PATH}"
```

File: prepare-image.sh [hunk 3]
```diff
@@ -2,75 +2,58 @@
...
+fi
 
-    python3 -m pip install --no-cache-dir --ignore-installed --prefix /usr -r "$IRONIC_PKG_LIST_FINAL" -c "${UPPER_CONSTRAINTS_PATH}"
+python3 -m pip install --no-cache-dir --ignore-installed --prefix /usr -r "$IRONIC_PKG_LIST_FINAL" -c "${UPPER_CONSTRAINTS_PATH}"
```
Comment (tuminoid) [inline prepare-image.sh:47]:
```suggestion
python3 -m pip install --no-cache-dir --ignore-installed --prefix /usr -r "${IRONIC_PKG_LIST_FINAL}" -c "${UPPER_CONSTRAINTS_PATH}"
```

File: prepare-image.sh [hunk 4]
```diff
@@ -2,75 +2,58 @@
...
-    curl -o /etc/yum.repos.d/delorean-deps.repo https://trunk.rdoproject.org/centos9-master/delorean-deps.repo
-
-    # NOTE(elfosardo): enable CRB repo for more python3 dependencies
-    dnf config-manager --set-enabled crb
-    dnf upgrade -y
-    xargs -rtd'\n' dnf install -y < "$IRONIC_PKG_LIST"
+# emulate uid/gid configuration to match rpm install
+IRONIC_UID=997
+IRONIC_GID=994
+BUILD_DEPS="python3-devel gcc git-core python3-setuptools python3-jinja2"
```
Comment (tuminoid) [inline prepare-image.sh:14]: If we are touching these lines now, let's use modern syntax and avoid shellcheck ignores later on. Also makes cleaner PRs going forward. Also sorted.

```suggestion
declare -a BUILD_DEPS=(
    gcc 
    git-core 
    python3-devel 
    python3-jinja2
    python3-setuptools 
)
```

File: prepare-image.sh [hunk 5]
```diff
@@ -2,75 +2,58 @@
...
+dnf upgrade -y
+# NOTE(dtantsur): pip is a requirement of python3 in CentOS
+# shellcheck disable=SC2086
+dnf install -y python3-pip $BUILD_DEPS
```
Comment (tuminoid) [inline prepare-image.sh:18]:
```suggestion
dnf install -y python3-pip "${BUILD_DEPS[@]}"
```
(only if using array when defining the list)

File: prepare-image.sh [hunk 6]
```diff
@@ -2,75 +2,58 @@
...
+getent group ironic > /dev/null || groupadd -r ironic -g "${IRONIC_GID}"
+getent passwd ironic > /dev/null || useradd -r -g ironic -u "${IRONIC_UID}" -s /sbin/nologin ironic -d /var/lib/ironic
 
-    # clean installed build dependencies
-    # shellcheck disable=SC2086
-    dnf remove -y $BUILD_DEPS
-fi
+# clean installed build dependencies
+# shellcheck disable=SC2086
+dnf remove -y $BUILD_DEPS
```
Comment (tuminoid) [inline prepare-image.sh:56]:
```suggestion
dnf remove -y "${BUILD_DEPS[@]}"
```
(only if using array when defining the list)

File: prepare-image.sh [hunk 7]
```diff
@@ -2,75 +2,58 @@
...
+mkdir -p /var/log/ironic /var/lib/ironic
+getent group ironic > /dev/null || groupadd -r ironic -g "${IRONIC_GID}"
+getent passwd ironic > /dev/null || useradd -r -g ironic -u "${IRONIC_UID}" -s /sbin/nologin ironic -d /var/lib/ironic
 
-    # clean installed build dependencies
-    # shellcheck disable=SC2086
-    dnf remove -y $BUILD_DEPS
-fi
+# clean installed build dependencies
+# shellcheck disable=SC2086
```
Comment (tuminoid) [inline prepare-image.sh:55]:
```suggestion
```
No longer needed with new syntax

File: prepare-image.sh [hunk 8]
```diff
@@ -2,75 +2,58 @@
...
-    dnf config-manager --set-enabled crb
-    dnf upgrade -y
-    xargs -rtd'\n' dnf install -y < "$IRONIC_PKG_LIST"
+# emulate uid/gid configuration to match rpm install
+IRONIC_UID=997
+IRONIC_GID=994
+BUILD_DEPS="python3-devel gcc git-core python3-setuptools python3-jinja2"
+dnf upgrade -y
+# NOTE(dtantsur): pip is a requirement of python3 in CentOS
+# shellcheck disable=SC2086
```
Comment (tuminoid) [inline prepare-image.sh:17]:
```suggestion
```
No longer needed with new syntax

**Review by tuminoid — COMMENTED** [review]
Suggestion: consistent curlys + use array for BUILD_DEPS for cleaner code, while touching those lines.

**Comment by elfosardo** [comment]
@tuminoid  thanks for the comments and suggestions, I didn't want to charge the PR too much so I just went for the minimum required changes, if that's ok for you I would apply your suggestions in a follow-up since it's passing CI

**Comment by dtantsur** [comment]
/approve

I agree with the comments. But since they concern the already existing code, I think it's a good idea to address them separately.

**Comment by tuminoid** [comment]
Follow up is ok as well.

**Pull request merged**